### PR TITLE
CACTUS-1054 Central Storybook

### PR DESCRIPTION
**Pull request by wilysword**
https://repayonline.atlassian.net/browse/CACTUS-1054

TL;DR Storybook is cool and I think it'll work just fine for what we want it to do, but `addon-docs` specifically may not be up to the challenge. If you read the whole description it might take even longer than reading the code changes 😜.

What I got done:
- Got the new storybook instance running.
- Found out the live-edit plugin linked in the ticket doesn't work (at least not without more config that I couldn't find on Google); _but_ our existing live-edit stuff does still work. We need to restyle it a bit so it fits in better with Storybook; also, in the long run I'd like something that's a bit more integrated with Storybook itself, so we can edit the actual stories and not just a completely separate example on the docs page, but for now maybe this is good enough?
- Switched a few stories over; showed a couple different ways of connecting the docs.
  - With `AccessibleField` I renamed the MDX file so that Storybook does some extra processing on it, and then included the result of that processing in the meta. Using that naming convention you can actually dispense with the `.tsx` file entirely, but I doubt we'll want to do that since it means all stories _must also_ appear inline in the docs page; including one or two as examples makes sense (code re-use and all), but except in very simple cases I doubt we'd want all of them.
  - With CheckBox I just imported the plain MDX, but for that to work I had to add some of the special processing from the other method myself. In some ways I like this better, but we'd definitely need a way to reduce that boilerplate, either by making our own webpack loader or a helper function or something.
  - You'll notice I changed the titles to something like `Cactus Web/Components/*`. That controls how Storybook sets up the hierarchy in the menu on the left, we could rename it to whatever. (Well, not _quite_ whatever: I wanted to have the first element be `@repay/cactus-web`, but there's no way to escape the `/`.)
- Got storyshots (sort of) working without needing to run a Storybook server: it can theoretically work just from the static build. This is where I ran out of time, so there are some problems left.
  - Jest and Storybook want the Babel config in different places. You can work around that, but you have to explicitly define the babel transform.
  - Storyshots doesn't like the MDX files. I tried adding Storybook's MDX loader, but it still wasn't happy.
  - All the other storyshots passed, but it didn't actually write any image files, not sure why.

Things I looked into and rejected:
- Storybook has a fair number of [test-related features](https://storybook.js.org/docs/react/writing-tests/test-runner) besides just storyshots. As far as I can tell, though, they're mostly just inferior versions of what jest & testing-library can do more naturally. It did give me an idea, though: importing stories for use in tests. It'd require some tweaks in some cases, but at least for basic tests the stories often cover the same use cases as the tests, so we could get a little code reuse in by rendering the stories in the tests then doing the assertions on that.

Future work:
- Getting the cactus theme addon updated to work in the new location.
- Honestly, maybe just replacing `addon-docs` entirely; at the least, replacing a fair bit of its functionality. It's a cool idea and works pretty well for anything basic or straightforward, but the instant you try to do anything beyond simple 1:1 control-to-prop stories, it breaks down.
  - `ArgsTable` is _not_ a good replacement for `PropsTable`. It has to be tied to a specific story, which means it won't show props that aren't controllable in that story. Even in simple cases like `Button`, that means it won't show things like `as` or `ref`, nor is there a way to specify that it supports all the normal `<button>` props.
  - And with the controls plugin, there's no way to to make it hide "uncontrollable" props like `onChange`; it'll show up in the Controls panel because it's in the args table, even though it's not a control. We probably need to just completely separate docs from controls.
  - The limitations on how the MDX files are loaded just aren't very friendly; like where I had to rename `docs/README.md` to `docs/README.story.mdx` to get it to show up. And that's not even getting into whether or not we can get Storyshots to work with MDX.
  - We might make our own set of components to use in MDX files.
  - Given how flaky both ArgsTable _and_ the PropsTable we use in Gatsby are as far as auto-detecting props, we may need to write our own definitions. Maybe something like a superset of both `propTypes` and `defaultProps`:
```
defineProps(Component, {
  propName: {
    type: somethingThatCanBeBothHumanReadableAndCanBeTransformedToAPropType,
    description: 'Sup',
    default: 'Hey',
    required: true,
  }
})
```
- Integrating live-edit into the stories more.
- Getting `addon-actions` to work. I removed it from our existing Storybook because it has really horrible formatting for event objects, but I do like the idea of actions showing up on the page instead of needing to open up the console. Maybe we could make our own addon that wraps their handlers with something that simplifies the event objects to where they're easier to read. (I'd probably want it to _also_ print to the console, still, because when you `console.log` an HTML element, hovering over it highlights the element on the page.)
- Fixing storyshots; there's also some accessibility stuff we might be able to add, either [with an addon](https://storybook.js.org/docs/react/writing-tests/accessibility-testing) or [directly alongside our storyshots](https://github.com/storybookjs/storybook/tree/next/code/addons/storyshots/storyshots-puppeteer#axetest).

Things I don't like:
- I actually really hate that it puts the controls in the URL as query parameters. I could see a use case for being _able_ to do that, but having them dynamically update along with the controls is just messy, and I hate that you can't turn it off.
- Can't really style it. The Storybook theming has a lot of options, but they all seem to be _very_ specific, so it's difficult to make broad changes. If this is going to be our main docs site, it'd be nice if we could make it a bit more REPAY-like. (Also the all-caps items in the left menu annoy me.)

### CODE REVIEW

**Comment by maxrcollins** [comment]
One of the things I wanted to make sure happened here is that we could have one set of docs for all of the modules. I'm not sure how difficult that would be.  If it's going to be a challenge, I'll make a separate ticket, but if not, could you add just a blank/placeholder page for Field.ts or the Icon page or something from a different module? Add it in a new `Cactus XXXX `section just below the `Cactus Web` section on the sidebar?

**Comment by wilysword** [comment]
@maxrcollins I didn't change the title, but the icon stories should already be in the new Storybook.

**Comment by maxrcollins** [comment]
@wilysword  Ohh I see it. It's all the way at the top. Nvm

**Comment by maxrcollins** [comment]
Initial thoughts: 

I'm completely happy with replacing addon-docs entirely. I think we should talk about this as a team because I also want to replace out cactus-theme add on. There has to be a better way to organize these.

As for how to load the components: I haven't played around with it much yet but why did you like the way of doing CheckBox better? I don't have a preference yet

**Comment by wilysword** [comment]
The CheckBox example is a little more straightforward in my opinion. Not on the code side, but that won't be applicable if we don't use `addon-docs`, and even if we do could be simplified with a helper function. I don't like having to duplicate the `Meta` information in both the MDX and JS files; I don't like having to name the MDX files `.story.mdx` (somehow that just seems a little messy to me). I'd prefer it if Storybook just automatically added the context for the current stories/component to the docs render function. Either way, though, we'd have to solve the problem I was having with MDX files breaking the storyshots.

> PropsTable

PropsTable is based on https://github.com/styleguidist/react-docgen-typescript; I think the auto-docs in Storybook are as well, actually. That works well for simple stuff, but styled-components really messes it up, especially when we layer styles on existing components (e.g. AccessibleField on top of FieldWrapper); it gets confused about what props should be included, and even what files the components belong in. It also sometimes just completely ignores some components, and I've never figured out the rhyme or reason there, since that one doesn't seem related to styled-components.

> Can't really style it.

I probably should be clear that I didn't look into it that much, focusing more on the docs & storyshots; I read the main docs pages about theming, but the Storybook docs are _far_ from comprehensive. There might be broader theming options, or a hook into the lower level library they use for styling.

**Comment by maxrcollins** [comment]
I'm having some es-lint errors on some of these new pages. It could be something with my linter but it looks like no? 
image

**Comment by Dhalton** [comment]
Not looking at the code but I was asked to provide some general feedback:

Really liking how this is shaping up so far. I think SB will be a better place to house our docs, and having our docs/testing/developer playground all in one place just seems like a no-brainer. Some thoughts on the things Glen brings up:

- I like how `addon-docs` organizes things with the tab-like view, but I see that it falls short and I like the idea of writing our own version of it. I assume it would be possible to keep the tab view the same way it is now. If so I think we should write our own addon but keep the tab layout.
- In that same-ish vein, @maxrcollins you mentioned replacing the cactus-theme addon (I think you're referring to the custom `cactus-addon` that we have to change theme values? I feel like we could use that as a base and build upon it; it would be nice imo if we could put it in the tab layout, so we'd have 3 tabs: story, docs, and theme. It could use some cleaning up, too. We could better describe what it is & what it's for, and make it look nicer. Just my ideas on what we should do with that piece of this.
- I _think_ Glen is suggesting that we define props manually and maintain them ourselves and if that's the case I think I can get behind that; although it would be nice to have them auto-generated, so far we've had lots of problems with that and things slip through the cracks and confuse our users. At least if we managed that ourselves we could make sure they're accurate and fully fleshed out instead of the crap the `PropsTable` yields.


Getting pulled away to fix a bug so I'll just leave this for now. Let me know if there's anything else specifically you want my feedback on

**Comment by wilysword** [comment]
These are the docs that talk about addon types: https://storybook.js.org/docs/react/addons/addon-types. Personally I think that theme would work best as a toolbar, rather than a panel (current) or tab (what Dhalton suggested). I'm not sure if it's possible, though; there seem to be some hints in the docs that it is, but all their examples are basically for simple drop-down-type data structures. If the choice is between panel and tab, though, I'd stick with panel; the theme settings are far more similar to "controls" than to the docs.

**Comment by maxrcollins** [comment]
> Not surprising since I didn't run the linter at all. Given the time-boxed nature of the ticket and the fact that there were always going to be followup tickets, I focused on getting it functional rather than appeasing the linter or Typescript.

I don't really want to merge something in with this many errors. I also don't want to make a followup ticket just to fix the linter errors from this ticket. Especially if so many of them are autofixable

**Comment by wilysword** [comment]
I suppose I can fix the linter errors, but I'm telling you right now I'm not touching Typescript. This is a _very_ early prototype, and will certainly be changing a great deal before it's released; there simply isn't a point spending hours of time fixing errors that will be overwritten later anyway.

File: storybook/.storybook/preview.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import { MINIMAL_VIEWPORTS } from '@storybook/addon-viewport'
+import { DocsContainer } from '@storybook/addon-docs'
+import React from 'react'
+
+import { StyleProvider } from '../../modules/cactus-web/src/'
+
+const STYLE = { global: true }
```
Comment (daniloPenaR) [inline storybook/.storybook/preview.js:7]: Is this value directly processed by Storybook? or we could get rid of it?

**Review by daniloPenaR — COMMENTED** [review]
A couple of initial comments on this, but this is looking good

File: modules/cactus-web/.babelrc [hunk 2]
```diff
@@ -1,8 +1 @@
-{
-  "presets": ["@repay/babel-preset"],
-  "env": {
-    "test": {
-      "plugins": ["require-context-hook"]
```
[the remark this answers is not in the document]
Comment (wilysword) [reply modules/cactus-web/.babelrc:5]: I have no idea what this is or what it's for; I don't even know if it _was_ ever needed. This installation has a `@storybook/babel-plugin-require-context-hook` package, which presumably makes the `require-context-hook` package obsolete or unnecessary, or maybe includes it. So I guess that's a long way of saying that either we don't need it at all, or Storybook is now including/configuring it for us.

File: modules/cactus-web/src/AccessibleField/AccessibleField.story.tsx [hunk 3]
```diff
@@ -1,22 +1,24 @@
 import React from 'react'
 
 import { AccessibleField } from '../'
-import { FIELD_ARGS, Story } from '../helpers/storybook'
+import { Story } from '../helpers/storybook'
+import docsMeta from './AccessibleField.story.mdx'
 
+delete docsMeta.includeStories
```
[the remark this answers is not in the document]
Comment (wilysword) [reply modules/cactus-web/src/AccessibleField/AccessibleField.story.tsx:7]: It makes it ignore the other stories in the file. This is because if you have a `story.mdx` file, they expect it to _replace_ the `story.tsx` file, not to be used alongside it.

File: modules/cactus-web/src/Select/Select.story.tsx [hunk 4]
```diff
@@ -62,7 +61,9 @@ export const CollisionsInLargeContainer: BasicStory = ({ showOptions, options, .
     </div>
   </React.Fragment>
 )
-CollisionsInLargeContainer.argTypes = { options: { control: 'array' }, margin: SPACE }
+CollisionsInLargeContainer.argTypes = { options: { control: 'array' }, margin: SPACE,
+    ...actions('onChange', 'onBlur', 'onFocus'),
+}
```
[the remark this answers is not in the document]
Comment (wilysword) [reply modules/cactus-web/src/Select/Select.story.tsx:66]: We had a custom addon doing a lot of stuff to fix the default behavior of controls in Storybook. Some of what the addon is doing will be irrelevant in the new instance, and some will need to be rewritten, so I didn't include it; that means some of the controls will be different than in the current Storybook, and some will just be broken.

File: storybook/.storybook/preview.js (hunk 1 above)
Comment (wilysword) [reply storybook/.storybook/preview.js:7]: Not needed, I was using it before but then I rewrote the code that was using it and forgot to remove it.

File: modules/cactus-icons/package.json [hunk 5]
```diff
@@ -14,7 +14,6 @@
   "sideEffects": false,
   "scripts": {
     "cleanup": "rm -rf ts i",
-    "start": "start-storybook -p 9001 -c .storybook",
     "dev": "echo -e \"This now builds in watch mode, use yarn icons start for storybook\\n\"; repay-scripts dev --lib --tree-shaking ts/index.ts",
```
[the remark this answers is not in the document]
Comment (wilysword) [reply modules/cactus-icons/package.json]: Yeah, I left them in on my first pass in case I needed to reference the commands being run (and was too lazy to look them up in the diff), then I just never went back for them.

**Review by daniloPenaR — COMMENTED** [review]
I like what I'm seeing at this point, Definitely I agree with the opinion that Storybook is a nice option to host our docs (way better than Gatsby). Regarding the ways for connecting the docs, what I don't like about `AccessibleField` is that `story.mdx` renaming, In the past months, I felt that Storybook did not have entire compatibility for this project, I mean when implementing controls a while ago you did a lot of work to fit that feature to the project's needs (correct me or compliment me if I am wrong), do you think that the _extra process_ storybook does to the `story.mdx` files currently fits the needs and the desires for this? 
On the other hand, I like the `CheckBox` approach the most. 
About the `theme` section as a toolbar panel, if possible, there should be nice, But for now, the possibility of a `Canvas/Docs/Theme` is more consistent for me.

File: modules/cactus-web/src/CheckBox/CheckBox.story.tsx [hunk 6]
```diff
@@ -1,17 +1,28 @@
...
+          mdxComponentAnnotations={meta}
+          mdxStoryNameToKey={{ 'Basic Usage': 'Basic Usage' }}
+        >
+          <DocsPage />
+        </AddContext>
+      ),
+    },
+  },
 } as const
+export default meta
```
[the remark this answers is not in the document]
Comment (wilysword) [reply modules/cactus-web/src/CheckBox/CheckBox.story.tsx:25]: They do the same thing, actually. I know I imported `Meta` in `CheckBox.mdx`, but I didn't need to because I didn't use it; I was just trying out some different things and then didn't clean up afterwards. `AccessibleField` has the `<Meta>` element because it does _not_ get that information from the story file, so instead you have to duplicate it in both places.

**Review by maxrcollins — APPROVED** [review]
This looks good to me. There's obviously going to need to be some cleanup but that can be done after we write our own add on and start filling out the components.

**Review by daniloPenaR — APPROVED** [review]
